servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherBuffer.java

```diff
@@ -0,0 +1,353 @@
+                            target.onNext(oldCollector.close());
+                        }
+                        return;
+                    }
+                }
+            }
+
+            @Override
+            public void onError(final Throwable t) {
+                terminateFromBoundariesSource(TerminalNotification.error(t));
+            }
+
+            @Override
+            public void onComplete() {
+                terminateFromBoundariesSource(TerminalNotification.complete());
+            }
+
+            void terminateFromItemsSource(final TerminalNotification terminalNotification) {
+                terminate(this, terminalNotification);
+            }
+
+            private void terminateFromBoundariesSource(final TerminalNotification terminalNotification) {
+                terminate(tCancellable, terminalNotification);
+            }
+
+            private void terminate(final Cancellable cancellable, final TerminalNotification terminalNotification) {
+                for (;;) {
+                    final Object cMaybeCollector = maybeCollector;
+                    if (cMaybeCollector == TERMINATED || cMaybeCollector instanceof DelayedTermination) {
+                        cancellable.cancel();
+                        return;
+                    }
+                    DelayedTermination delayedTermination =
+                            new DelayedTermination(cMaybeCollector, terminalNotification, cancellable);
+                    if (maybeCollectorUpdater.compareAndSet(
+                            BufferSubscriber.this, cMaybeCollector, delayedTermination)) {
+                        // If we raced with ADDING then hand-off termination to ADDING
+                        if (cMaybeCollector != ADDING) {
+                            tryTerminate(delayedTermination);
+                        }
+                        break;
+                    }
+                }
+            }
+
+            void tryTerminate(final DelayedTermination delayedTermination) {
+                for (;;) {
+                    final long cPendingBuffers = pendingBuffers;
+                    if (cPendingBuffers <= 0) {
+                        return;
+                    }
+                    if (pendingBuffersUpdater.compareAndSet(
+                            BufferSubscriber.this, cPendingBuffers, cPendingBuffers - 1)) {
+                        for (;;) {
+                            final Object cMaybeCollector = maybeCollector;
+                            if (cMaybeCollector == TERMINATED) {
+                                return;
+                            }
+                            assert cMaybeCollector == delayedTermination;
+                            if (maybeCollectorUpdater.compareAndSet(
+                                    BufferSubscriber.this, delayedTermination, TERMINATED)) {
+                                delayedTermination.terminate(target);
```

Comment:
IIUC we may concurrently invoke methods on `target`

```
T1 - BoundariesSubscriber#onNext(..)
  target.onNext(..)
T2 - BufferSubscriber#onError(..)
  terminateFromItemsSource(..)
  cMaybeCollector = maybeCollector -> Collector
  maybeCollectorUpdater.CAS(cMaybeCollector, DelayedTermination) -> true
  tryTerminate()
  cPendingBuffers = pendingBuffers -> 100 (>0)
  pendingBuffersUpdater.CAS(100, 99) -> true
  cMaybeCollector = maybeCollector -> DelayedTermination
  target.onError(..)
  // target.onError and target.onNext concurrently invoked
```

Reply:
good catch; lemme fix